### PR TITLE
feature: add support to ``clarify.py`` for time series explainability jobs 

### PR DESCRIPTION
**Pull request by rvasahu-amazon**
A successor to [this](https://github.com/aws/sagemaker-python-sdk/pull/4214) PR.

*Description of changes:*

- Added ``TimeSeriesDataConfig``, ``TimeSeriesModelConfig``, and ``AsymmetricShapleyValueConfig`` (child of ``ExplainabilityConfig``) classes. Added unit tests for these new components.
- Modified ``DataConfig``, and ``ModelConfig`` to take ``TimeSeriesDataConfig`` and ``TimeSeriesModelConfig`` respectively as parameters. Added unit tests for these changes.
- Modified `` _AnalysisConfigGenerator`` to support the modified ``DataConfig`` and ``ModelConfig`` plus the new ``AsymmetricShapleyValueConfig``. Added unit tests for these changes.

*Testing done:*

Clarify unit test results: 
```
============================================= 145 passed in 1.02s ==============================================
```

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 640dbd11b4d5db4c2241c3569b00c5c83ff27c9a
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=02753be5-27c3-4bd7-b2ff-319dd88a5f2f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/clarify.py [hunk 1]
```diff
@@ -664,6 +946,38 @@ def get_config(self):
...
+        self,
+        forecast: str,
+    ):
+        """Initializes model configuration fields for TimeSeries explainability use cases.
+
+        Args:
+            forecast (str): JMESPath expression to extract the forecast result.
+
+        Raises:
+            AssertionError: when either ``forecast`` or ``forecast_horizon`` are not provided
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:962]: `forecast_horizon` not used, right?

File: src/sagemaker/clarify.py [hunk 2]
```diff
@@ -326,6 +388,99 @@ class DatasetType(Enum):
...
+
+    ```
+    item_id="ids"
+    timestamp="timestamps"
+    target_time_series="target_ts"
+    related_time_series=["rts1", "rts2"]
+    static_covariates=["scv1", "scv2"]
+    ```
+
+    Below is an example ``ITEM_RECORDS`` dataset for time series explainability:
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:419]: do we support item records or not? if we do, let's extend the schema accordingly, please

```
            SchemaOptional("dataset_format"): And(
                str,
                Use(str.lower),
                lambda s: s
                in (
                    "columns",
                    "timestamp_records",
                ),
            ),
```

File: src/sagemaker/clarify.py [hunk 3]
```diff
@@ -430,6 +693,11 @@ def __init__(
                 Note: For JSON, the JMESPath query must result in a list of labels for each
                 sample.  For JSON Lines, it must result in the label for each line.
                 Only a single label per sample is supported at this time.
+            headers (str): List of column names in the dataset. If not provided, Clarify will
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:696]: should it be list of strings (type annotation)? 
```suggestion
            headers ([str]): List of column names in the dataset. If not provided, Clarify will
```

File: src/sagemaker/clarify.py (hunk 2 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:419]: Good catch, will do (internally as well).

File: src/sagemaker/clarify.py (hunk 1 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:962]: Yes, will remove.

File: src/sagemaker/clarify.py [hunk 4]
```diff
@@ -85,6 +98,22 @@
         SchemaOptional("excluded_columns"): [Or(int, str)],
         SchemaOptional("joinsource_name_or_index"): Or(str, int),
         SchemaOptional("group_variable"): Or(str, int),
+        SchemaOptional("time_series_data_config"): {
+            "target_time_series": Or(str, int),
+            "item_id": Or(str, int),
+            "timestamp": Or(str, int),
+            SchemaOptional("related_time_series"): Or([str], [int]),
+            SchemaOptional("static_covariates"): Or([str], [int]),
+            SchemaOptional("dataset_format"): And(
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:107]: Why is it optional?

File: src/sagemaker/clarify.py [hunk 5]
```diff
@@ -1400,6 +1735,88 @@ def get_explainability_config(self):
...
+        Args:
+            direction (str): Type of explanation to be used. Available explanation
+                types are ``"chronological"``, ``"anti_chronological"``, and ``"bidirectional"``.
+            granularity (str): Explanation granularity to be used. Available granularity options
+                are ``"timewise"`` and ``"fine_grained"``.
+            num_samples (None or int): Number of samples to be used in the Asymmetric Shapley
+                Value forecasting algorithm. Only applicable when using ``"fine_grained"``
+                explanations.
+            baseline (str or dict): Link to a baseline configuration or a dictionary for it.
+            # TODO: improve above.
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:1779]: Let's improve

File: src/sagemaker/clarify.py [hunk 6]
```diff
@@ -1400,6 +1735,88 @@ def get_explainability_config(self):
...
+            ), f"{direction} and {granularity} granularity are not supported together."
+        elif num_samples:  # validate num_samples is not provided when unnecessary
+            raise ValueError("``num_samples`` is only used for fine-grained explanations.")
+        # set explanation type and (if provided) num_samples in internal config dictionary
+        _set(direction, "direction", self.asymmetric_shapley_value_config)
+        _set(granularity, "granularity", self.asymmetric_shapley_value_config)
+        _set(
+            num_samples, "num_samples", self.asymmetric_shapley_value_config
+        )  # _set() does nothing if a given argument is None
+        # TODO: add sdk-side validation to baseline
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:1812]: Let's improve

File: src/sagemaker/clarify.py [hunk 7]
```diff
@@ -1400,6 +1735,88 @@ def get_explainability_config(self):
...
+
+        Args:
+            direction (str): Type of explanation to be used. Available explanation
+                types are ``"chronological"``, ``"anti_chronological"``, and ``"bidirectional"``.
+            granularity (str): Explanation granularity to be used. Available granularity options
+                are ``"timewise"`` and ``"fine_grained"``.
+            num_samples (None or int): Number of samples to be used in the Asymmetric Shapley
+                Value forecasting algorithm. Only applicable when using ``"fine_grained"``
+                explanations.
+            baseline (str or dict): Link to a baseline configuration or a dictionary for it.
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:1778]: Do we support this for time-series explainability: `Link to a baseline configuration`? Did we test it?

File: src/sagemaker/clarify.py [hunk 8]
```diff
@@ -278,6 +307,36 @@
...
+                    str,
+                    Use(str.lower),
+                    lambda s: s
+                    in (
+                        "timewise",
+                        "fine_grained",
+                    ),
+                ),
+                SchemaOptional("num_samples"): int,
+                SchemaOptional("baseline"): Or(
```
Comment (aws-byeldos) [inline src/sagemaker/clarify.py:331]: Why is baseline optional? What would be a default value for related_time-series?

File: src/sagemaker/clarify.py (hunk 4 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:107]: As discussed offline, we are leaving it defined in schema as optional for now.

File: src/sagemaker/clarify.py (hunk 3 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:696]: Yes, will fix.

File: src/sagemaker/clarify.py (hunk 5 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:1779]: Will do.

File: src/sagemaker/clarify.py (hunk 6 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py:1812]: Will do.

File: src/sagemaker/clarify.py (hunk 7 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py]: To my knowledge we support this but going to test this and will come back to you.

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: ef9659f215a14a4b8c0cda4fcfb37d1bfe9f39f8
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=be5053a1-aed6-4806-b5d6-05fcd951127b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/clarify.py (hunk 7 above)
Comment (rvasahu-amazon) [reply src/sagemaker/clarify.py]: We support this.

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: bbeba313c40ce5e173eb730de7bdc0ecbbf8d74a
* Result: SUCCEEDED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4c1d63f2-a980-47bd-ab8f-016e0d1cfc25%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: a8ad1d1bdece5f760be9ae8bfa9a0e644bc71c0f
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=76f29de5-a49b-44fa-8319-8fa0d558fda0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by aws-byeldos — APPROVED** [review]

**Comment by aws-byeldos** [comment]
let's improve the validation, so it checks if baseline is S3 uri

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: b535f530ab1211b38a63f81750fc4b4072a5af4b
* Result: SUCCEEDED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=baaa8586-3d18-4229-abf3-ccc6402855ab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by zicanl-amazon — APPROVED** [review]

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 222fb759c0d4fa1c1b201ef111375bafc2c398c8
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=8d7c91f5-2124-46c7-96f3-c183f29ec9c4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: ba8ac77cf52eb006485cdeb4997be0c95618c069
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0ac5a914-9e9a-4376-b7d8-6aa6c53558d0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by aws-byeldos — APPROVED** [review]

**Review by zicanl-amazon — APPROVED** [review]

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: ff581815ff9c387b1967c18331c7bd4a6e68a315
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=65ee2e7e-e10c-4752-aa7c-1d4fffaf39f1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 6274f5fde3ce398b4432ffd20012f8a28a49d3e3
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=de5d4abf-6110-421a-91b4-294039c6cf25%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 6274f5fde3ce398b4432ffd20012f8a28a49d3e3
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=cbb3de2b-4b0b-4759-9260-5449fd2a5585%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 89e6100d304f0142955d741fb1e5df5a54270a3b
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=cc6697d7-6252-4eec-bc9f-ff24f43c03cd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by benieric — APPROVED** [review]

**Pull request merged**